hikari/impl/special_endpoints.py

```diff
@@ -700,6 +713,82 @@ async def __anext__(self) -> audit_logs.AuditLog:
         return log
 
 
+class GuildThreadIterator(iterators.BufferedLazyIterator[_GuildThreadChannelT]):
+    """Iterator implemented for guild thread endpoints."""
+
+    __slots__: typing.Sequence[str] = (
+        "_before_is_timestamp",
+        "_deserialize",
+        "_entity_factory",
+        "_has_more",
+        "_next_before",
+        "_request_call",
+        "_route",
+    )
+
+    def __init__(
+        self,
+        deserialize: _ThreadDeserailzeSig[_GuildThreadChannelT],
+        entity_factory: entity_factory_.EntityFactory,
+        request_call: _RequestCallSig,
+        route: routes.CompiledRoute,
+        before: undefined.UndefinedOr[str],
+        before_is_timestamp: bool,
+    ) -> None:
+        super().__init__()
+        self._before_is_timestamp = before_is_timestamp
+        self._deserialize = deserialize
+        self._entity_factory = entity_factory
+        self._has_more = True
+        self._next_before = before
+        self._request_call = request_call
+        self._route = route
+
+    async def _next_chunk(self) -> typing.Optional[typing.Generator[_GuildThreadChannelT, typing.Any, None]]:
+        if not self._has_more:
+            return None
+
+        query = data_binding.StringMapBuilder()
+        query.put("limit", 100)
+        query.put("before", self._next_before)
+
+        response = await self._request_call(compiled_route=self._route, query=query)
+        assert isinstance(response, dict)
+
+        if not (threads := response["threads"]):
+            # Since GET is idempotent, has_more should always be false if there
+            # are no threads in the current response.
+            self._has_more = False
+            return None
+
+        self._has_more = response["has_more"]
+        members = {int(member["id"]): member for member in response["members"]}
+        if self._before_is_timestamp:
+            self._next_before = min(
```

Comment:
discord guarantees a returned order for both cases here so this can just rely on that instead